File: MATTHEW/leetcode/AIcoding/4.GPT-4o/go_solutions/LCR 037-XagZNi.go
func asteroidCollision(asteroids []int) []int {
    stack := []int{}

    for _, asteroid := range asteroids {
        alive := true
        for alive && asteroid < 0 && len(stack) > 0 && stack[len(stack)-1] > 0 {
            if stack[len(stack)-1] < -asteroid {
                stack = stack[:len(stack)-1]
            } else if stack[len(stack)-1] == -asteroid {
                stack = stack[:len(stack)-1]
                alive = false
            } else {
                alive = false
            }
        }
        if alive {
            stack = append(stack, asteroid)
        }
    }

    return stack
}

File: MATTHEW/leetcode/AIcoding/4.GPT-4o/rust_solutions/LCR 037-XagZNi.rs
pub fn asteroid_collision(asteroids: Vec<i32>) -> Vec<i32> {
    let mut stack = Vec::new();
    for &asteroid in asteroids.iter() {
        let mut alive = true;
        while alive && asteroid < 0 && !stack.is_empty() && *stack.last().unwrap() > 0 {
            let top = stack.pop().unwrap();
            if top == -asteroid {
                alive = false;
            } else if top > -asteroid {
                stack.push(top);
                alive = false;
            } else {
                continue;
            }
        }
        if alive {
            stack.push(asteroid);
        }
    }
    stack
}